src/wallet/wallet.cpp

```diff
@@ -81,13 +81,33 @@ std::shared_ptr<CWallet> GetWallet(const std::string& name)
     return nullptr;
 }
 
+static std::mutex g_wallet_release_mutex;
+static std::condition_variable g_wallet_release_cond;
+
+void WaitForWalletRelease(std::shared_ptr<CWallet>& wallet)
+{
+    std::weak_ptr<CWallet> weak_wallet(wallet);
+    wallet.reset();
+
+    std::unique_lock<std::mutex> lk(g_wallet_release_mutex);
+    while (!weak_wallet.expired()) {
```

Comment:
It looks like this might not be sufficient if we want to guarantee that the wallet is really destroyed. From https://en.cppreference.com/w/cpp/memory/weak_ptr/expired:

> The destructor for the managed object may not yet have been called, but this object's destruction is imminent (or may have already happened).

A more reliable but cumbersome approach might be to add wallet members:

```c++
std::promise<void> m_unload_promise;
std:shared_future<void> m_unload_future = m_unload_promise.get_future();
```

and call `m_unloaded_promise.set_value()` in the destructor, and then `future.wait()` here.